src/action.rs

```diff
@@ -0,0 +1,405 @@
+    )>,
+    bindings: Query<&Binding>,
+) {
+    let (action_bindings, settings, fns, mut value, mut state, mut events, mut time) =
+        actions.get_mut(trigger.target()).unwrap();
+
+    *time = Default::default();
+    events.set_if_neq(ActionEvents::new(*state, ActionState::None));
+    state.set_if_neq(Default::default());
+    value.set_if_neq(ActionValue::zero(value.dim()));
+
+    fns.trigger(
+        &mut commands,
+        trigger.target(),
+        *state,
+        *events,
+        *value,
+        *time,
+    );
+
+    if let Some(action_bindings) = action_bindings
+        && settings.require_reset
+    {
+        pending.extend(bindings.iter_many(action_bindings).copied());
+    }
+}
+
+/// Component that represents a user action.
+///
+/// Entities with this component needs to be spawned with [`ActionOf<C>`]
+/// relationship in order to be evaluated.
+///
+/// Holds value defined by [`InputAction::Output`].
+///
+/// See the required components for other data associated with the action
+/// that can be accessed without static typing.
+#[derive(Component, Deref, DerefMut)]
+#[require(
+    Name::new(any::type_name::<A>()),
+    ActionFns::new::<A>(),
+    ActionValue::zero(A::Output::DIM),
+    ActionSettings,
+    ActionState,
+    ActionEvents,
+    ActionTime,
+)]
+pub struct Action<A: InputAction>(A::Output);
+
+impl<A: InputAction> Clone for Action<A> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<A: InputAction> Copy for Action<A> {}
+
+impl<A: InputAction> Default for Action<A> {
+    fn default() -> Self {
+        Self(Default::default())
+    }
+}
+
+impl<A: InputAction> Action<A> {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+/// Marker for a gameplay-related action.
+///
+/// Used together with [`Action<C>`] and [`events`] to statically define the type.
+///
+/// To implement the trait you can use the [`InputAction`](bevy_enhanced_input_macros::InputAction)
+/// derive to reduce boilerplate. Just specify `action_output` attribute with the type.
+///
+/// ```
+/// # use bevy::prelude::*;
+/// # use bevy_enhanced_input::prelude::*;
+/// #[derive(InputAction)]
+/// #[action_output(Vec2)]
+/// struct Move;
+/// ```
+pub trait InputAction: 'static {
+    /// What type of value this action will output.
+    ///
+    /// - Use [`bool`] for button-like actions (e.g., `Jump`).
+    /// - Use [`f32`] for single-axis actions (e.g., `Zoom`).
+    /// - For multi-axis actions, like `Move`, use [`Vec2`] or [`Vec3`].
+    type Output: ActionOutput;
+}
+
+/// Marks a type which can be used as [`InputAction::Output`].
+pub trait ActionOutput: Into<ActionValue> + Default + Send + Sync + Debug + Clone + Copy {
+    /// Dimension of this output.
+    const DIM: ActionValueDim;
+
+    /// Converts the value into the action output type.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the value represents a different type.
+    fn unwrap_value(value: ActionValue) -> Self;
```

Comment:
Non-blocking, but I don't think this should panic. Instead, return a result and let the caller unwrapper or recover.

Reply:
Same as above - users won't use it and I used panic only because this should never happen: we create right dynamic type for an action and just unwrap it into the right static type. If the panic triggers - it's a but in my implementation :slightly_smiling_face: